File: TauResearch/ViewModels/VideoRecordingViewModel.swift
// VideoRecordingViewModel.swift
// Mirrors VideoRecordingViewModel.kt — wires CameraController + RecordingService + GPS gating
// + detection smoothing.

import Foundation
import AVFoundation
import CoreLocation
import SwiftData
import Observation

struct RecordingSegment: Equatable {
    let videoPath: String
    let gpxPath: String
    let createdAt: Date
}

@Observable
@MainActor
final class VideoRecordingViewModel {

    // MARK: - Public state mirroring the Compose VM

    var uiText: String = "Idle"
    var recording: Bool = false
    var previewRunning: Bool = false
    var paused: Bool = false
    var recordingDuration: TimeInterval = 0
    var detectionResult: DetectionResult?
    var lastSegment: RecordingSegment?
    var lastVideoPath: String?

    /// Mirrors `recordingVm.location` from the Android VM. Reads from
    /// `RecordingService.shared.currentLocation`; the service is `@Observable`, so
    /// SwiftUI re-renders whenever a new fix arrives.
    var location: CLLocation? { service.currentLocation }

    var recordingReady: Bool = false
    var gpsReadyForRecording: Bool = false
    var selectedDynamicRange: Int = 1  // 1 = standard, matches the Android constant

    // MARK: - Dependencies

    let controller: CameraController
    private let service = RecordingService.shared
    private let repo: RecordingRepository

    // MARK: - Internal state

    private var pipelineReady = false
    private var gpsConsecutiveCount = 0
    private var gpsGateActive = false
    private var durationTimer: Task<Void, Never>?
    private var gpsTimeoutTask: Task<Void, Never>?
    private let requiredGpsPointsForRecording = 10
    private let gpsAcquireTimeoutSeconds: TimeInterval = 20

    // Smoothing state — matches the Kotlin alpha = 0.15
    private let alpha: Float = 0.15
    private var smoothedConfidence: Float = 0
    private var smoothedObjectLuma: Float = 0
    private var smoothedFrameLuma: Float = 0

    init(context: ModelContext) {
        self.controller = CameraController.shared
        self.repo = RecordingRepository(context: context)
        wireDetectionHandler()
    }

    private func wireDetectionHandler() {
        controller.runKiwiInferenceEnabled = useKiwiFruitMetering()
        controller.onDetectionUpdate = { [weak self] result in
            self?.handleDetection(result)
        }
    }

    /// Fruit scan? Read the cached scan-type id and look up its name in the fetched catalogue.
    func useKiwiFruitMetering() -> Bool {
        guard let submitted = cachedSiteBlock(),
              let json = ApiCache.shared.getFetchedSiteBlock(),
              let data = json.data(using: .utf8),
              let fetched = try? JSONDecoder.tau.decode(SiteBlockResponse.self, from: data) else {
            return false
        }
        let name = fetched.scan_type.first(where: { $0.id == submitted.scanTypeId })?.name?.trimmingCharacters(in: .whitespaces)
        return name?.lowercased() == "fruit"
    }

    // MARK: - Preview / Recording

    func startPreview(device: AVCaptureDevice, fps: Int, iso: Int, exposureTimeNs: Int64, onStatus: @escaping (String) -> Void) {
        Task {
            // Kick the GPS stream BEFORE we start the gate. Previously the gate polled
            // `service.currentLocation` while `LocationManager.startUpdatingLocation`
            // was still idle, so the gate timed out and the Start-Recording button
            // never lit up.
            service.startLocationUpdates()

            await controller.startPreview(device: device,
                                           targetFps: fps,
                                           iso: iso,
                                           exposureTimeNs: exposureTimeNs) { msg in
                self.uiText = msg
                onStatus(msg)
            }
            previewRunning = true
            recording = false
            startGpsGateForPreview()
            await prepareRecordingPipeline()
        }
    }

    func stopCamera() {
        let wasRecording = recording
        controller.stop()
        previewRunning = false
        recording = false
        // Don't cut the GPX off mid-track. If a recording was active when stopCamera
        // was called (e.g., scene backgrounded), leave the GPS stream running so the
        // recording's existing logger can finalize cleanly; otherwise stop the stream
        // to save battery.
        if !wasRecording {
            service.stopLocationUpdates()
        }
        resetRecordingReadinessState()
    }

    func startRecording(device: AVCaptureDevice, fps: Int, iso: Int, exposureTimeNs: Int64) {
        guard recordingReady else {
            uiText = !gpsReadyForRecording ? "Waiting for GPS…" : "Preparing camera pipeline. Please wait…"
            return
        }
        uiText = "Starting recording…"
        let scanType = resolveScanTypeLabel()
        let path = service.startRecording(
            device: device,
            targetFps: fps,
            iso: iso,
            exposureTimeNs: exposureTimeNs,
            analysisSize: controller.defaultAnalysisSize,
            scanType: scanType,
            onStatus: { msg in self.uiText = msg }
        )
        lastVideoPath = path
        recording = true
        paused = false
        previewRunning = true
        gpsGateActive = false
        gpsTimeoutTask?.cancel()
        startDurationTimer()
    }

    func stopRecording() {
        let videoPath = service.stopRecording()
        let gpxPath = service.stopGpxLogging()
        _ = service.stopCsvLogging()

        recording = false
        paused = false
        resetRecordingReadinessState()
        stopDurationTimer()
        controller.stop()
        previewRunning = false
        // Recording fully done — stop the GPS stream so it doesn't drain battery while
        // the user looks at the "Ready to Upload" card.
        service.stopLocationUpdates()

        if let videoPath {
            let gpx = gpxPath ?? (videoPath as NSString).deletingPathExtension + ".gpx"
            lastSegment = RecordingSegment(videoPath: videoPath, gpxPath: gpx, createdAt: .now)
            uiText = "Video + GPX Saved ✅"
            if let block = cachedSiteBlock() {
                Task { await repo.saveRecording(videoPath: videoPath, gpxPath: gpx, siteBlock: block) }
            }
        } else {
            uiText = "Stopped recording."
        }
    }

    func togglePauseResume() {
        guard recording else { return }
        if !paused {
            if controller.pauseRecording() {
                paused = true
                uiText = "Paused ⏸️"
                durationTimer?.cancel()
            } else {
                uiText = "Pause not supported ❌"
            }
        } else {
            if controller.resumeRecording() {
                paused = false
                uiText = "Recording resumed ▶️"
                startDurationTimer(isResume: true)
            } else {
                uiText = "Resume failed ❌"
            }
        }
    }

    func logMetrics(
        countPerFrame: Int,
        currentBrightness: Int,
        targetedBrightness: Int,
        adjustedBrightness: Int,
        currentIso: Int,
        adjustedIso: Int,
        currentShutterDenom: Int,
        adjustedShutterDenom: Int
    ) {
        guard recording else { return }
        service.appendMetricsRow(
            countPerFrame: countPerFrame,
            currentBrightness: currentBrightness,
            targetedBrightness: targetedBrightness,
            adjustedBrightness: adjustedBrightness,
            currentIso: currentIso,
            adjustedIso: adjustedIso,
            currentShutterDenom: currentShutterDenom,
            adjustedShutterDenom: adjustedShutterDenom
        )
    }

    // MARK: - Detection smoothing

    private func handleDetection(_ new: DetectionResult) {
        smoothedConfidence = alpha * new.confidenceScore + (1 - alpha) * smoothedConfidence
        if new.objectLuma > 0 {
            if smoothedObjectLuma == 0 { smoothedObjectLuma = Float(new.objectLuma) }
            smoothedObjectLuma = alpha * Float(new.objectLuma) + (1 - alpha) * smoothedObjectLuma
        } else {
            smoothedObjectLuma = 0
        }
        if new.avgFrameLuma > 0 {
            if smoothedFrameLuma == 0 { smoothedFrameLuma = Float(new.avgFrameLuma) }
            smoothedFrameLuma = alpha * Float(new.avgFrameLuma) + (1 - alpha) * smoothedFrameLuma
        } else {
            smoothedFrameLuma = 0
        }
        var smoothed = new
        smoothed.confidenceScore = smoothedConfidence
        smoothed.objectLuma = Int(smoothedObjectLuma)
        smoothed.avgFrameLuma = Int(smoothedFrameLuma)
        detectionResult = smoothed
    }

    // MARK: - GPS gate

    private func startGpsGateForPreview() {
        gpsTimeoutTask?.cancel()
        gpsConsecutiveCount = 0
        gpsGateActive = true
        gpsReadyForRecording = false
        recomputeRecordingReady()
        gpsTimeoutTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(20 * 1_000_000_000))
            guard let self else { return }
            if previewRunning, !recording, !gpsReadyForRecording {
                stopCamera()
                uiText = "GPS data was not found."
            }
        }
        // Mirror Android: poll currentLocation each second.
        Task { [weak self] in
            while let self, self.gpsGateActive, !self.recording {
                self.handleGpsGate(location: self.service.currentLocation)
                try? await Task.sleep(nanoseconds: 500_000_000)
            }
        }
    }

    private func handleGpsGate(location: CLLocation?) {
        guard let location, location.isValidCoordinate else {
            gpsConsecutiveCount = 0
            return
        }
        gpsConsecutiveCount += 1
        if gpsConsecutiveCount >= requiredGpsPointsForRecording {
            gpsReadyForRecording = true
            gpsGateActive = false
            gpsTimeoutTask?.cancel()
            recomputeRecordingReady()
        }
    }

    private func recomputeRecordingReady() {
        recordingReady = pipelineReady && gpsReadyForRecording
    }

    private func resetRecordingReadinessState() {
        pipelineReady = false
        gpsReadyForRecording = false
        recordingReady = false
        gpsConsecutiveCount = 0
        gpsGateActive = false
        gpsTimeoutTask?.cancel()
        gpsTimeoutTask = nil
    }

    // MARK: - Pipeline prep

    private func prepareRecordingPipeline() async {
        pipelineReady = false
        recomputeRecordingReady()
        let scanType = resolveScanTypeLabel()
        let artifactsReady = service.prepareRecordingArtifacts(scanType: scanType) != nil
        let detectionReady = await controller.warmUpDetection()
        pipelineReady = artifactsReady && detectionReady
        recomputeRecordingReady()
        if !pipelineReady, previewRunning {
            uiText = "Warm-up failed. Retry preview."
        }
    }

    // MARK: - Duration

    private func startDurationTimer(isResume: Bool = false) {
        if !isResume { recordingDuration = 0 }
        durationTimer?.cancel()
        durationTimer = Task { [weak self] in
            while !(Task.isCancelled) {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                guard let self else { return }
                if Task.isCancelled { return }
                self.recordingDuration += 1
            }
        }
    }

    private func stopDurationTimer() {
        durationTimer?.cancel()
        durationTimer = nil
    }

    // MARK: - Cache helpers

    private func cachedSiteBlock() -> SiteBlockCreateRequest? {
        guard let json = ApiCache.shared.getSubmitSiteBlock(),
              let data = json.data(using: .utf8) else { return nil }
        return try? JSONDecoder.tau.decode(SiteBlockCreateRequest.self, from: data)
    }

    private func resolveScanTypeLabel() -> String {
        guard let submitted = cachedSiteBlock() else { return "" }
        guard let json = ApiCache.shared.getFetchedSiteBlock(),
              let data = json.data(using: .utf8),
              let fetched = try? JSONDecoder.tau.decode(SiteBlockResponse.self, from: data) else {
            return String(submitted.scanTypeId)
        }
        return fetched.scan_type.first(where: { $0.id == submitted.scanTypeId })?.name ?? String(submitted.scanTypeId)
    }
}

private extension CLLocation {
    var isValidCoordinate: Bool {
        let lat = coordinate.latitude
        let lon = coordinate.longitude
        return (-90...90).contains(lat) && (-180...180).contains(lon)
    }
}

File: app/src/main/java/com/example/cameraaccess/viewmodel/VideoRecordingViewModel.kt
package com.example.cameraaccess.viewmodel

import android.app.Application
import android.content.*
import android.content.pm.PackageManager
import android.graphics.Rect
import android.location.Location
import android.os.IBinder
import android.util.Log
import android.util.Range
import android.util.Size
import androidx.activity.result.ActivityResultLauncher
import androidx.compose.runtime.mutableStateOf
import androidx.core.content.ContextCompat
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.example.cameraaccess.controller.Camera2Controller
import com.example.cameraaccess.data.model.ApiCache
import com.example.cameraaccess.data.model.SiteBlockCreateRequest
import com.example.cameraaccess.data.model.SiteBlockResponse
import com.example.cameraaccess.data.repositories.RecordingRepository
import com.example.cameraaccess.processing.DetectionResult
import com.example.cameraaccess.service.RecordingService
import com.example.cameraaccess.ui.record.RecordingSegment
import com.google.gson.Gson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class VideoRecordingViewModel(
    initialController: Camera2Controller,
    private val context: Context,
    application: Application
) : AndroidViewModel(application) {

    private var _controller: Camera2Controller = initialController
    val controller: Camera2Controller get() = _controller

    val uiText = mutableStateOf("Idle")
    val recording = mutableStateOf(false)
    val previewRunning = mutableStateOf(false)
    val paused = mutableStateOf(false)

    private val _recordingDuration = MutableStateFlow(0L)
    val recordingDuration: StateFlow<Long> = _recordingDuration
    private var durationJob: Job? = null

    var lastVideoPath: String? = null
        private set

    private val _lastSegment = MutableStateFlow<RecordingSegment?>(null)
    val lastSegment: StateFlow<RecordingSegment?> = _lastSegment
    
    private val _location = MutableStateFlow<Location?>(null)
    val location: StateFlow<Location?> = _location

    private val _detectionResult = MutableStateFlow<DetectionResult?>(null)
    val detectionResult: StateFlow<DetectionResult?> = _detectionResult
    private val _recordingReady = MutableStateFlow(false)
    val recordingReady: StateFlow<Boolean> = _recordingReady
    private val _gpsReadyForRecording = MutableStateFlow(false)
    val gpsReadyForRecording: StateFlow<Boolean> = _gpsReadyForRecording
    private var pipelineReady = false
    private var gpsConsecutiveCount = 0
    private var gpsGateActive = false
    private var gpsTimeoutJob: Job? = null

    // Smoothing state
    private var smoothedConfidence = 0f
    private var smoothedObjectLuma = 0f
    private var smoothedFrameLuma = 0f
    private val alpha = 0.15f // Smoothing factor

    private val repository = RecordingRepository(application)
    private val cache = ApiCache(application)
    private val gson = Gson()

    private var recordingService: RecordingService? = null
    private var isBound = false
    private var warmupJob: Job? = null
    private var lastPreviewVideoSize: Size? = null
    private var lastPreviewFpsRange: Range<Int>? = null
    private var lastPreviewDynamicRange: Long = Camera2Controller.DYNAMIC_RANGE_STANDARD

    private val TAG = "VideoRecordingViewModel"
    private val requiredGpsPointsForRecording = 10
    private val gpsAcquireTimeoutMs = 20_000L

    val selectedDynamicRange = MutableStateFlow(1L) // Default to STANDARD

    private val serviceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            val binder = service as RecordingService.LocalBinder
            recordingService = binder.getService()
            _controller = recordingService!!.cameraController
            isBound = true
            Log.d(TAG, "Service Connected")

            applyKiwiInferenceSetting(_controller)
            _controller.onDetectionUpdate = { result ->
                processAndSmoothResult(result)
            }

            viewModelScope.launch {
                recordingService?.currentLocation?.collect { loc ->
                    _location.value = loc
                    handleGpsGateLocationUpdate(loc)
                }
            }

            if (previewRunning.value && !recording.value) {
                pipelineReady = false
                recomputeRecordingReady()
                val warmedVideoSize = lastPreviewVideoSize
                if (warmedVideoSize != null) {
                    startGpsGateForPreview()
                    prepareRecordingPipeline(
                        videoSize = warmedVideoSize,
                        fpsRange = lastPreviewFpsRange,
                        dynamicRangeProfile = lastPreviewDynamicRange
                    )
                }
            }
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            recordingService = null
            isBound = false
            Log.d(TAG, "Service Disconnected")
        }
    }

    private fun processAndSmoothResult(newResult: DetectionResult) {
        // Smooth the Confidence Score
        smoothedConfidence = (alpha * newResult.confidenceScore) + ((1 - alpha) * smoothedConfidence)
        
        // Smooth the Object Luma (only if identified)
        if (newResult.objectLuma > 0) {
            if (smoothedObjectLuma == 0f) smoothedObjectLuma = newResult.objectLuma.toFloat()
            smoothedObjectLuma = (alpha * newResult.objectLuma) + ((1 - alpha) * smoothedObjectLuma)
        } else {
            smoothedObjectLuma = 0f
        }

        if (newResult.avgFrameLuma > 0) {
            if (smoothedFrameLuma == 0f) smoothedFrameLuma = newResult.avgFrameLuma.toFloat()
            smoothedFrameLuma = (alpha * newResult.avgFrameLuma) + ((1 - alpha) * smoothedFrameLuma)
        } else {
            smoothedFrameLuma = 0f
        }

        _detectionResult.value = newResult.copy(
            confidenceScore = smoothedConfidence,
            objectLuma = smoothedObjectLuma.toInt(),
            avgFrameLuma = smoothedFrameLuma.toInt(),
            frameWidth = newResult.frameWidth,
            frameHeight = newResult.frameHeight
        )
    }

    init {
        Intent(application, RecordingService::class.java).also { intent ->
            application.bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
        }

        applyKiwiInferenceSetting(_controller)
        _controller.onDetectionUpdate = { result ->
            processAndSmoothResult(result)
        }
    }

    /**
     * Fruit scan runs kiwi detection and can meter exposure on detected fruit; other scan types
     * meter the full frame only ([Camera2Controller.runKiwiInferenceEnabled] false).
     */
    fun useKiwiFruitMetering(): Boolean = isFruitScanTypeSelected()

    private fun isFruitScanTypeSelected(): Boolean {
        val submitted = getCachedSiteBlock() ?: return false
        val id = submitted.scanTypeId
        val fetchedJson = cache.getFetchedSiteBlock() ?: return false
        return try {
            val fetched = gson.fromJson(fetchedJson, SiteBlockResponse::class.java)
            val name = fetched.scan_type.firstOrNull { it.id == id }?.name?.trim()
            "Fruit".equals(name, ignoreCase = true)
        } catch (e: Exception) {
            Log.w(TAG, "Error resolving scan type for kiwi metering", e)
            false
        }
    }

    private fun applyKiwiInferenceSetting(controller: Camera2Controller) {
        controller.runKiwiInferenceEnabled = isFruitScanTypeSelected()
    }

    private fun hasPermission(permission: String) =
        ContextCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED

    fun startPreview(
        cameraId: String,
        previewSize: Size,
        videoSize: Size,
        fpsRange: Range<Int>?,
        iso: Int,
        exposureTimeNs: Long,
        dynamicRangeProfile: Long,
        onStatus: (String) -> Unit
    ) {
        val serviceIntent = Intent(context, RecordingService::class.java).apply {
            action = RecordingService.ACTION_START_PREVIEW
        }
        try {
            ContextCompat.startForegroundService(context, serviceIntent)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start foreground service for preview", e)
        }

        controller.startPreview(
            cameraId = cameraId,
            previewSize = previewSize,
            fpsRange = fpsRange,
            iso = iso,
            exposureTimeNs = exposureTimeNs,
            dynamicRangeProfile = dynamicRangeProfile,
            onStatus = { msg ->
                uiText.value = msg
                onStatus(msg)
            }
        )
        previewRunning.value = true
        recording.value = false
        startGpsGateForPreview()
        lastPreviewVideoSize = videoSize
        lastPreviewFpsRange = fpsRange
        lastPreviewDynamicRange = dynamicRangeProfile
        prepareRecordingPipeline(videoSize, fpsRange, dynamicRangeProfile)
    }

    fun stopCamera() {
        controller.stop()
        previewRunning.value = false
        recording.value = false
        resetRecordingReadinessState()
        warmupJob?.cancel()
        lastPreviewVideoSize = null

        val serviceIntent = Intent(context, RecordingService::class.java).apply {
            action = RecordingService.ACTION_STOP_SERVICE
        }
        context.startService(serviceIntent)
    }

    fun startRecording(
        permissionLauncher: ActivityResultLauncher<Array<String>>,
        cameraId: String,
        previewSize: Size,
        videoSize: Size,
        fpsRange: Range<Int>?,
        iso: Int,
        exposureTimeNs: Long
    ) {
        if (!_recordingReady.value) {
            uiText.value = if (!_gpsReadyForRecording.value) {
                "Waiting for GPS..."
            } else {
                "Preparing camera pipeline. Please wait..."
            }
            return
        }
        val permissions = mutableListOf(
            android.Manifest.permission.CAMERA,
            android.Manifest.permission.RECORD_AUDIO,
            android.Manifest.permission.ACCESS_FINE_LOCATION
        )
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
            permissions.add(android.Manifest.permission.POST_NOTIFICATIONS)
        }

        if (permissions.any { !hasPermission(it) }) {
            uiText.value = "Missing permissions ❌"
            permissionLauncher.launch(permissions.toTypedArray())
            return
        }

        val serviceIntent = Intent(context, RecordingService::class.java).apply {
            action = RecordingService.ACTION_START_RECORDING
        }
        try {
            ContextCompat.startForegroundService(context, serviceIntent)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start foreground service", e)
            uiText.value = "Service error: ${e.message}"
            return
        }

        uiText.value = "Starting recording…"

        // Dual pipeline: 720p 10-bit YCBCR_P010 for green detection; encoder gets full HDR/10-bit.
        val analysisSize = controller.defaultAnalysisSize

        val scanTypeLabel = resolveScanTypeLabel()

        val videoPath = recordingService?.startRecording(
            cameraId = cameraId,
            previewSize = previewSize,
            videoSize = videoSize,
            fpsRange = fpsRange,
            iso = iso,
            exposureTimeNs = exposureTimeNs,
            dynamicRangeProfile = selectedDynamicRange.value,
            analysisSize = analysisSize,
            scanType = scanTypeLabel
        ) { msg -> uiText.value = msg } ?: controller.startRecording(
            cameraId, previewSize, videoSize, fpsRange, iso, exposureTimeNs,
            selectedDynamicRange.value, analysisSize
        ) { msg -> uiText.value = msg }

        lastVideoPath = videoPath
        recording.value = true
        paused.value = false
        previewRunning.value = true
        gpsGateActive = false
        gpsTimeoutJob?.cancel()
        startDurationTimer()
    }

    fun stopRecording() {
        val videoPath = recordingService?.stopRecording() ?: controller.stopRecording()
        val gpxPath = recordingService?.stopGpxLogging()

        recording.value = false
        paused.value = false
        resetRecordingReadinessState()
        stopDurationTimer()

        controller.stop()
        previewRunning.value = false

        val serviceIntent = Intent(context, RecordingService::class.java).apply {
            action = RecordingService.ACTION_STOP_SERVICE
        }
        context.startService(serviceIntent)

        if (videoPath != null) {
             val finalGpxPath = gpxPath ?: videoPath.replace(".mp4", ".gpx")

            _lastSegment.value = RecordingSegment(videoPath = videoPath, gpxPath = finalGpxPath)
            uiText.value = "Video + GPX Saved ✅"
            getCachedSiteBlock()?.let {
                viewModelScope.launch {
                    repository.saveRecording(videoPath, finalGpxPath, it)
                }
            }
        } else {
            uiText.value = "Stopped recording."
        }
    }

    private fun getCachedSiteBlock(): SiteBlockCreateRequest? {
        return try {
            cache.getSubmitSiteBlock()?.let {
                gson.fromJson(it, SiteBlockCreateRequest::class.java)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing cached site block", e)
            null
        }
    }

    /**
     * Returns a human-readable scan-type label (name from the fetched site-block catalogue,
     * falling back to the numeric id, or empty string if nothing is cached).
     */
    private fun resolveScanTypeLabel(): String {
        val submitted = getCachedSiteBlock() ?: return ""
        val id = submitted.scanTypeId
        val fetchedJson = cache.getFetchedSiteBlock() ?: return id.toString()
        return try {
            val fetched = gson.fromJson(fetchedJson, SiteBlockResponse::class.java)
            fetched.scan_type.firstOrNull { it.id == id }?.name ?: id.toString()
        } catch (e: Exception) {
            Log.w(TAG, "Error parsing fetched site block for scan type", e)
            id.toString()
        }
    }

    private fun prepareRecordingPipeline(
        videoSize: Size,
        fpsRange: Range<Int>?,
        dynamicRangeProfile: Long
    ) {
        warmupJob?.cancel()
        pipelineReady = false
        recomputeRecordingReady()
        warmupJob = viewModelScope.launch {
            val scanTypeLabel = resolveScanTypeLabel()
            val artifactsReady = withContext(Dispatchers.IO) {
                recordingService?.prepareRecordingArtifacts(scanTypeLabel) != null
            }
            val detectionReady = withContext(Dispatchers.Default) { controller.warmUpDetection() }
            val recorderReady = withContext(Dispatchers.Default) {
                controller.preWarmRecorder(videoSize, fpsRange, dynamicRangeProfile)
            }
            pipelineReady = artifactsReady && detectionReady && recorderReady
            recomputeRecordingReady()
            if (!pipelineReady && previewRunning.value) {
                uiText.value = "Warm-up failed. Retry preview."
            }
        }
    }

    private fun startGpsGateForPreview() {
        gpsTimeoutJob?.cancel()
        gpsConsecutiveCount = 0
        gpsGateActive = true
        _gpsReadyForRecording.value = false
        recomputeRecordingReady()
        gpsTimeoutJob = viewModelScope.launch {
            delay(gpsAcquireTimeoutMs)
            if (previewRunning.value && !recording.value && !_gpsReadyForRecording.value) {
                stopCamera()
                uiText.value = "GPS data was not found."
            }
        }
    }

    private fun handleGpsGateLocationUpdate(location: Location?) {
        if (!gpsGateActive || recording.value) return
        if (location == null || !location.isValidCoordinate()) {
            gpsConsecutiveCount = 0
            return
        }
        gpsConsecutiveCount += 1
        if (gpsConsecutiveCount >= requiredGpsPointsForRecording) {
            _gpsReadyForRecording.value = true
            gpsGateActive = false
            gpsTimeoutJob?.cancel()
            recomputeRecordingReady()
        }
    }

    private fun recomputeRecordingReady() {
        _recordingReady.value = pipelineReady && _gpsReadyForRecording.value
    }

    private fun resetRecordingReadinessState() {
        pipelineReady = false
        _gpsReadyForRecording.value = false
        _recordingReady.value = false
        gpsConsecutiveCount = 0
        gpsGateActive = false
        gpsTimeoutJob?.cancel()
        gpsTimeoutJob = null
    }

    /**
     * Appends a per-frame metrics row to the CSV that accompanies the current recording.
     * No-op when no recording is in progress or the service isn't bound.
     */
    fun logMetrics(
        countPerFrame: Int,
        currentBrightness: Int,
        targetedBrightness: Int,
        adjustedBrightness: Int,
        currentIso: Int,
        adjustedIso: Int,
        currentShutterDenom: Int,
        adjustedShutterDenom: Int
    ) {
        if (!recording.value) return
        recordingService?.appendMetricsRow(
            countPerFrame,
            currentBrightness,
            targetedBrightness,
            adjustedBrightness,
            currentIso,
            adjustedIso,
            currentShutterDenom,
            adjustedShutterDenom
        )
    }

    fun togglePauseResume() {
        if (!recording.value) return
        if (!paused.value) {
            if (controller.pauseRecording()) {
                paused.value = true
                uiText.value = "Paused ⏸️"
                durationJob?.cancel()
            } else {
                uiText.value = "Pause not supported ❌"
            }
        } else {
            if (controller.resumeRecording()) {
                paused.value = false
                uiText.value = "Recording resumed ▶️"
                startDurationTimer(isResume = true)
            } else {
                uiText.value = "Resume failed ❌"
            }
        }
    }

    private fun startDurationTimer(isResume: Boolean = false) {
        if (!isResume) {
            _recordingDuration.value = 0L
        }
        durationJob?.cancel()
        durationJob = viewModelScope.launch {
            while (true) {
                delay(1000)
                _recordingDuration.value += 1000
            }
        }
    }

    private fun stopDurationTimer() {
        durationJob?.cancel()
        durationJob = null
    }

    override fun onCleared() {
        super.onCleared()
        warmupJob?.cancel()
        gpsTimeoutJob?.cancel()
        try {
            if (isBound) {
                getApplication<Application>().unbindService(serviceConnection)
                isBound = false
            }
        } catch (e: Exception) {
            Log.w(TAG, "Error unbinding service in onCleared", e)
        }
    }
}

private fun Location.isValidCoordinate(): Boolean {
    val lat = latitude
    val lon = longitude
    return lat in -90.0..90.0 && lon in -180.0..180.0
}
